src/js/media-tooltip.ts

```diff
@@ -0,0 +1,239 @@
+ *
+ * @cssproperty --media-font - `font` shorthand property.
+ * @cssproperty --media-font-weight - `font-weight` property.
+ * @cssproperty --media-font-family - `font-family` property.
+ * @cssproperty --media-font-size - `font-size` property.
+ * @cssproperty --media-text-content-height - `line-height` of button text.
+ *
+ * @cssproperty --media-tooltip-background - `background` color
+ * @cssproperty --media-tooltip-display - `display` of tooltip
+ * @cssproperty --media-tooltip-z-index - `z-index` of tooltip
+ * @cssproperty --media-tooltip-padding - `padding` of tooltip
+ * @cssproperty --media-tooltip-border-radius - `border-radius` of tooltip
+ * @cssproperty --media-tooltip-filter - `filter` property of tooltip, for drop-shadow
+ * @cssproperty --media-tooltip-white-space - `white-space` property of tooltip
+ * @cssproperty --media-tooltip-arrow-display - `display` property of tooltip arrow
+ */
+class MediaTooltip extends globalThis.HTMLElement {
+  static get observedAttributes(): string[] {
+    return [Attributes.POSITION, Attributes.CONTAINER];
+  }
+
+  arrowEl: HTMLElement;
+
+  constructor() {
+    super();
+
+    if (!this.shadowRoot) {
+      // Set up the Shadow DOM if not using Declarative Shadow DOM.
+      this.attachShadow({ mode: 'open' });
+      this.shadowRoot.appendChild(template.content.cloneNode(true));
+    }
+
+    this.arrowEl = this.shadowRoot.querySelector('#arrow');
+  }
+
+  // Adjusts tooltip position relative to the closest specified container
+  // such that it doesn't spill out of the left or right sides. Only applies
+  // to 'top' and 'bottom' positioned tooltips.
+  updateXOffset = () => {
+    const position = this.position;
+
+    // we don't offset against tooltips coming out of left and right sides
+    if (position === 'left' || position === 'right') {
+      // could have been offset before switching to a new position
+      this.style.removeProperty('--media-tooltip-offset-x');
+      return;
+    }
+
+    const containingSelector = this.container;
+    const containingEl = closestComposedNode(this, containingSelector);
+    if (!containingEl) return;
+    const { x: containerX, width: containerWidth } =
+      containingEl.getBoundingClientRect();
+    const { x: tooltipX, width: tooltipWidth } = this.getBoundingClientRect();
+    const tooltipRight = tooltipX + tooltipWidth;
+    const containerRight = containerX + containerWidth;
+    const offsetXVal = this.style.getPropertyValue('--media-tooltip-offset-x');
+    const currOffsetX = offsetXVal
+      ? parseFloat(offsetXVal.replace('px', ''))
+      : 0;
+
+    // we might have already offset the tooltip previously so we remove it's
+    // current offset from our calculations
+    const leftDiff = tooltipX - containerX + currOffsetX;
+    const rightDiff = tooltipRight - containerRight + currOffsetX;
+
+    // out of left bounds
+    if (leftDiff < 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${leftDiff}px`);
```

Comment:
How this is all working is hard to understand from scratch so don't be shy with code comments. Maybe a summary of how `media-tooltip-offset-x` is ultimately used to set both sides.

Reply:
Have added some more comments to this that that hopefully explains this better. The only difference between needing to be brought in from the left or from the right is if you need a positive offset (push to the right) or negative offset (pull it further left) so it just re-uses the variable for both cases, setting a positive or negative offset depending on what needs to be done.